<launch>

  <arg name="frame_id"/>

  <arg name="map_size_x_"/>
  <arg name="map_size_y_"/>
  <arg name="map_size_z_"/>


  <arg name="map_origin_x_"/>
  <arg name="map_origin_y_"/>
  <arg name="map_origin_z_"/>


  <arg name="box_min_x"/>
  <arg name="box_min_y"/>
  <arg name="box_min_z"/>
  <arg name="box_max_x"/>
  <arg name="box_max_y"/>
  <arg name="box_max_z"/>

  <arg name="odometry_topic"/>
  <arg name="sensor_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>

  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>

  <arg name="max_vel"/>
  <arg name="max_acc"/>

  <arg name="pose_type"/>
  <arg name="waypoints_topic"/>
  <arg name="tracker_topic"/>
  <arg name="srv_name"/>


  <!-- main node -->
  <!-- <node pkg="exploration_manager" name="exploration_node" type="exploration_node" output="screen" launch-prefix="valgrind"> -->
  <node pkg="exploration_manager" name="exploration_node" type="exploration_node" output="screen" >
    <remap from ="/odom_world" to="$(arg odometry_topic)"/>
    <remap from ="/map_ros/pose"  to = "$(arg sensor_pose_topic)"/> 
    <remap from ="/map_ros/depth" to = "$(arg depth_topic)"/>
    <remap from ="/map_ros/cloud" to="$(arg cloud_topic)"/>

    <!-- yuwei -->
    <remap from="~waypoints"  to="$(arg waypoints_topic)"/>
    <remap from="~tracker_cmd"  to="$(arg tracker_topic)"/>
    <param name="srv_name"     value="$(arg srv_name)"/>
    <param name="plan_frame_id"  value ="$(arg frame_id)"/>

    <param name="sdf_map/resolution"      value="0.1" /> 
    <param name="sdf_map/map_size_x"   value="$(arg map_size_x_)" /> 
    <param name="sdf_map/map_size_y"   value="$(arg map_size_y_)" /> 
    <param name="sdf_map/map_size_z"   value="$(arg map_size_z_)" /> 

    <param name="sdf_map/map_origin_x"   value="$(arg map_origin_x_)" /> 
    <param name="sdf_map/map_origin_y"   value="$(arg map_origin_y_)" /> 
    <param name="sdf_map/map_origin_z"   value="$(arg map_origin_z_)" /> 

    <!-- Global map inflation -->
    <param name="sdf_map/inflation_xy"     value="0.15" /> 
    <param name="sdf_map/inflation_z"     value="0.15" /> 

    <!-- Local map inflation -->
    <param name="sdf_map/inflate_local" value="false" type="bool"/> 
    <param name="sdf_map/obstacles_inflation"     value="0.15" /> 
    <param name="sdf_map/local_bound_inflate"    value="0.5"/>
    <param name="sdf_map/local_map_margin"      value="50"/>
    <!-- <param name="sdf_map/obstacles_inflation"     value="0.099" />  -->
    <param name="sdf_map/ground_height"        value="-0.1"/>
    <param name="sdf_map/default_dist"        value="0.0"/>
    
    <param name="sdf_map/p_hit"  value="0.70"/>
    <param name="sdf_map/p_miss" value="0.35"/>
    <param name="sdf_map/p_min"  value="0.12"/>
    <param name="sdf_map/p_max"  value="0.97"/>
    <param name="sdf_map/p_occ"  value="0.80"/>
    <param name="sdf_map/p_pred_occ"  value="0.65"/>
    <param name="sdf_map/min_ray_length" value="0.3"/>
    <param name="sdf_map/max_ray_length" value="5.0"/>
    <param name="sdf_map/virtual_ceil_height"   value="-10"/>
    <param name="sdf_map/optimistic" value="false" type="bool"/>
    <param name="sdf_map/signed_dist" value="false" type="bool"/>
    <param name="sdf_map/box_min_x" value="$(arg box_min_x)" type="double"/>
    <param name="sdf_map/box_min_y" value="$(arg box_min_y)" type="double"/>
    <param name="sdf_map/box_min_z" value="$(arg box_min_z)" type="double"/>
    <param name="sdf_map/box_max_x" value="$(arg box_max_x)" type="double"/>
    <param name="sdf_map/box_max_y" value="$(arg box_max_y)" type="double"/>
    <param name="sdf_map/box_max_z" value="$(arg box_max_z)" type="double"/>

    <param name="map_ros/cx" value="$(arg cx)"/>
    <param name="map_ros/cy" value="$(arg cy)"/>
    <param name="map_ros/fx" value="$(arg fx)"/>
    <param name="map_ros/fy" value="$(arg fy)"/>
    <param name="map_ros/depth_filter_maxdist"   value="5.0"/>
    <param name="map_ros/depth_filter_mindist"   value="0.2"/>
    <param name="map_ros/depth_filter_margin"    value="2"/>
    <param name="map_ros/k_depth_scaling_factor" value="1000.0"/>
    <param name="map_ros/skip_pixel" value="2"/>
    <param name="map_ros/esdf_slice_height" value="0.3"/>
    <param name="map_ros/visualization_truncate_height"   value="10.09"/>
    <param name="map_ros/visualization_truncate_low"   value="-2.0"/>
    <param name="map_ros/show_occ_time"  value="false"/>
    <param name="map_ros/show_esdf_time" value="false"/>
    <param name="map_ros/show_all_map" value="true"/>
    <param name="map_ros/frame_id"      value="world"/>
    <param name="map_ros/pose_type"     value="$(arg pose_type)"/>

    <!-- Fsm -->
    <param name="fsm/thresh_replan1" value="0.8" type="double"/>
    <param name="fsm/thresh_replan2" value="2.0" type="double"/>
    <param name="fsm/thresh_replan3" value="1.5" type="double"/>
    <param name="fsm/estop_time"     value="0.8" type="double"/>
    <!-- <param name="fsm/thresh_replan1" value="0.00001" type="double"/> -->
    <!-- <param name="fsm/thresh_replan2" value="100.0" type="double"/> -->
    <!-- <param name="fsm/thresh_replan3" value="0.1" type="double"/> -->
    <param name="fsm/replan_time" value="0.3" type="double"/>

    <!-- Exploration manager -->
    <param name="exploration/refined_radius" value="5.0" type="double"/>
    <param name="exploration/max_decay" value="0.8" type="double"/>
    <param name="exploration/top_view_num" value="15" type="int"/>
    <param name="exploration/vm" value="$(eval 1.0 * arg('max_vel'))" type="double"/>
    <param name="exploration/am" value="$(eval 1.0 * arg('max_acc'))" type="double"/>
    <param name="exploration/yd" value="$(eval 60 * 3.1415926 / 180.0)" type="double"/>
    <param name="exploration/ydd" value="$(eval 90 * 3.1415926 / 180.0)" type="double"/>
    <param name="exploration/w_dir" value="1.5" type="double"/>
    <param name="exploration/relax_time" value="1.0" type="double"/>
    <param name="exploration/detect_semantics" value="true" type="bool"/>
    
    <param name="exploration/radius_far"    value="5.0" type="double"/>
    <param name="exploration/radius_close"  value="3.0" type="double"/>
    <param name="exploration/radius_noplan" value="0.5" type="double"/>

    <param name="exploration/terminate_condition_vol_enabled" value="true" type="bool"/>
    <param name="exploration/terminate_condition_vol" value="248.0" type="double"/>
    
    <param name="frontier/exp_3d" value="false" type="bool"/>
    <param name="frontier/cluster_min" value="100" type="int"/>
    <param name="frontier/cluster_size_xy" value="1.2" type="double"/>
    <param name="frontier/cluster_size_z" value="3.0" type="double"/>
    <param name="frontier/room_min_candidate_dist" value="0.20" type="double"/>
    <param name="frontier/min_candidate_dist" value="0.75" type="double"/>
    <param name="frontier/min_candidate_obs_clearance" value="0.3" type="double"/>
    <param name="frontier/min_candidate_clearance" value="0.4" type="double"/>

    <param name="frontier/candidate_dphi" value="$(eval 15 * 3.1415926 / 180.0)" type="double"/>
    <param name="frontier/candidate_rnum" value="3" type="int"/>
    <param name="frontier/candidate_rmin" value="0.5" type="double"/>
    <param name="frontier/candidate_rmax" value="1.2" type="double"/>
    <!-- <param name="frontier/candidate_rnum" value="3" type="int"/> -->
    <!-- <param name="frontier/candidate_rmin" value="1.5" type="double"/> -->
    <!-- <param name="frontier/candidate_rmax" value="2.5" type="double"/> -->
    <param name="frontier/down_sample" value="3" type="int"/>;
    <param name="frontier/min_visib_num" value="12" type="int"/>;
    <param name="frontier/min_info_gain" value="500" type="int"/>;
    <param name="frontier/min_view_finish_fraction" value="0.2" type="double"/>;
    <param name="frontier/h_fov" value="1.51844" type="double"/>;
    <param name="frontier/v_fov" value="1.01229" type="double"/>;
    <param name="frontier/min_ray_length" value="0.3" type="double"/>;
    <param name="frontier/max_ray_length" value="5.0" type="double"/>;
    <remap from="frontier/occ_map/pred" to="/frontier/occ_map/pred" />
    <param name="frontier/pred_size_x" value="80" type="int"/>;
    <param name="frontier/pred_size_y" value="80" type="int"/>;
    <param name="frontier/pred_size_z" value="24" type="int"/>;

    <!-- Door detection related param -->
    <param name="frontier/door_confirm_thres" value="1.2" type="double"/>;
    <param name="frontier/door_cross_thres" value="1.2" type="double"/>;
    <param name="frontier/door_exit_thres" value="1.2" type="double"/>;
    <param name="frontier/door_confirm_dis" value="1.0" type="double"/>;
    <param name="frontier/door_cross_dis" value="2.0" type="double"/>;
    <param name="frontier/door_exit_dis" value="1.0" type="double"/>;
    <param name="frontier/min_vote" value="5" type="int"/>; 
    <param name="frontier/h_min_line_length" value="3" type="double"/>;
    <param name="frontier/h_max_line_gap" value="5" type="double"/>;
    <param name="frontier/v_min_line_length" value="10" type="double"/>;
    <param name="frontier/v_max_line_gap" value="6" type="double"/>;
    <param name="frontier/h_min_door" value="9" type="double"/>;
    <param name="frontier/h_max_door" value="16" type="double"/>;
    <param name="frontier/v_min_door" value="9" type="double"/>;
    <param name="frontier/v_max_door" value="16" type="double"/>;

    <!-- Perception utils -->
    <param name="perception_utils/top_angle" value="0.56125" type="double"/>;
    <param name="perception_utils/left_angle" value="0.69222" type="double"/>;
    <param name="perception_utils/right_angle" value="0.68901" type="double"/>;
    <param name="perception_utils/max_dist" value="3.0" type="double"/>;
    <param name="perception_utils/vis_dist" value="1.0" type="double"/>;


    <param name="heading_planner/yaw_diff" value="$(eval 30 * 3.1415926 / 180.0)" type="double"/>
    <param name="heading_planner/half_vert_num" value="5" type="int"/>
    <param name="heading_planner/lambda1" value="2.0" type="double"/>
    <param name="heading_planner/lambda2" value="1.0" type="double"/>
    <param name="heading_planner/max_yaw_rate" value="$(eval 10 * 3.1415926 / 180.0)" type="double"/>
    <param name="heading_planner/w" value="20000.0" type="double"/>
    <param name="heading_planner/weight_type" value="1" type="double"/>

    <!-- planner manager -->
    <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/max_dyaw" value="$(eval 60 * 3.1415926 / 180.0)" type="double"/>
    <param name="manager/max_ddyaw" value="$(eval 90 * 3.1415926 / 180.0)" type="double"/>

    <!-- kinodynamic path searching -->
    <param name="manager/use_astar_path" value="true" type="bool"/>
    <param name="manager/use_kino_acc_path" value="true" type="bool"/>
    <param name="manager/min_time" value="true" type="bool"/>
  
    <param name="manager/map_size_x"   value="$(arg map_size_x_)" /> 
    <param name="manager/map_size_y"   value="$(arg map_size_y_)" /> 

    <!-- planning ceilings -->
    <param name="manager/map_size_z"   value="2.2" /> 
    <param name="manager/map_origin_x"   value="$(arg map_origin_x_)" /> 
    <param name="manager/map_origin_y"   value="$(arg map_origin_y_)" /> 
    <param name="manager/map_origin_z"   value="$(arg map_origin_z_)" /> 
    <param name="manager/time_res"   value="0.2" /> 


    <!-- kinodynamic path searching -->
    <param name="search/max_tau" value="0.6" type="double"/>
    <param name="search/init_max_tau" value="0.6" type="double"/>
    <param name="search/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="search/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="search/max_jerk" value="4" type="double"/>
    <param name="search/w_time" value="3.0" type="double"/>
    <param name="search/horizon" value="8.0" type="double"/>
    <param name="search/lambda_heu" value="5.0" type="double"/>
    <param name="search/resolution_astar" value="0.1" type="double"/>
    <param name="search/margin" value="0.2" type="double"/>
    <param name="search/allocate_num" value="100000" type="int"/>
    <param name="search/optimistic" value="false" type="bool"/>
    <param name="search/vis_check_num" value="15" type="int"/>

    <param name="search/w_uyaw" value="3.0" type="double"/>
    <param name="search/w_semantics" value="1.0" type="double"/>
    <param name="search/w_info" value="0.004" type="double"/>
    <param name="search/yaw_resolution"  value="0.1" type="double"/>
    <param name="search/max_search_time" value="0.03" type="double"/>

    <param name="astar/lambda_heu" value="2.0" type="double"/>
    <param name="astar/resolution_astar" value="0.2" type="double"/>
    <param name="astar/allocate_num" value="1000000" type="int"/>
    <param name="astar/max_search_time" value="0.05" type="double"/>

    <!-- optimization -->
    <param name="optimization/w_time"     value="0.5" type="double"/>
    <param name="optimization/w_vel"      value="10000.0" type="double"/>
    <param name="optimization/w_acc"      value="10000.0" type="double"/>
    <param name="optimization/w_sta_obs"  value="20.0" type="double"/>

    <param name="optimization/w_yaw_time"     value="1.0" type="double"/>
    <param name="optimization/w_refyaw"       value="40.0" type="double"/>
    <param name="optimization/w_dyaw"         value="2000.0" type="double"/>
    <param name="optimization/w_ddyaw"        value="4000.0" type="double"/>

  </node>

</launch>
